opencl/btx_clinterval_callbacks.cpp

```diff
@@ -0,0 +1,354 @@
+                                               int64_t ts, const char *event_class_name,
+                                               const char *hostname, int64_t vpid,
+                                               uint64_t vtid,
+                                               cl_command_queue command_queue,
+                                               cl_int errcode_ret_val) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  auto entry_device = state->entry_state.get_data<thapi_device_id>(key);
+  const auto root_device = state->device_to_root_device[
+                                              hp_device_t(hostname, vpid, entry_device)];
+  hp_command_queue_t cq_key{hostname, vpid, command_queue};
+  if (root_device != 0) {
+    state->command_queue_to_device[cq_key] = dsd_t(root_device, entry_device) ;
+  } else {
+    state->command_queue_to_device[cq_key] = dsd_t(entry_device, 0) ;
+  }
+}
+
+static void kernel_info_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                 const char *event_class_name, const char *hostname,
+                                 int64_t vpid, uint64_t vtid, cl_kernel kernel,
+                                 char *function_name) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  // TODO: do we need to save a copy?
+  state->kernel_to_name[hp_kernel_t(hostname, vpid, kernel)] = function_name;
+}
+
+static void launch_kernel_entry_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                         const char *event_class_name,
+                                         const char *hostname, int64_t vpid,
+                                         uint64_t vtid, cl_command_queue command_queue,
+                                         cl_kernel kernel) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  auto name = state->kernel_to_name[hp_kernel_t(hostname, vpid, kernel)];
+  state->profiled_function_name_and_ts[key] = fn_ts_t(name, ts);
+  state->function_name_to_dsd[hpt_function_name_t(hostname, vpid, vtid, name)] =
+    state->command_queue_to_device[hp_command_queue_t(hostname, vpid, command_queue)];
+}
+
+static void launch_function_entry_callback(void *btx_handle, void *usr_data, int64_t ts,
+                                           const char *event_class_name,
+                                           const char *hostname, int64_t vpid,
+                                           uint64_t vtid,
+                                           cl_command_queue command_queue) {
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key{hostname, vpid, vtid};
+  auto name_str = strip_event_class_name_entry(event_class_name);
+  const char *name = name_str.c_str();
+  state->profiled_function_name_and_ts[key] = fn_ts_t(name, ts);
+  state->function_name_to_dsd[hpt_function_name_t(hostname, vpid, vtid, name)] =
```

Comment:
Remove type, and should we keep `name` as a string?

Reply:
I like idea of keeping name as string, then we don't have to worry about c ptr going dest going out of scope.

Reply:
One issue with this, is all the compound types defined in xprof_utils, using char * for function name (e.g. hpt_function_name_t). But I guess if we limit using the name to only in the struct declaration, is less of an issue